horizomer/utils/tree.py

```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env python
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2015--, The Horizomer Development Team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+#
+# functions relevant to phylogenetic tree operations
+#
+
+from skbio import TreeNode
+
+
+def compare_topology(node1, node2):
+    """Test whether the topologies of two trees are identical.
+
+    Parameters
+    ----------
+    tree1 : skbio.TreeNode
+        first tree in comparison
+    tree2 : skbio.TreeNode
+        second tree in comparison
+
+    Returns
+    -------
+    bool
+        whether the topologies are identical
+
+    Notes
+    -----
+    Flipping child nodes does not change topology.
+    Branch lengths are ignored.
+    """
+    if len(node1.children) != len(node2.children):
+        return False
+    elif len(node1.children) > 0:
+        # check if sorted child names are identical
+        children1 = sorted(node1.children, key=lambda x: x.name)
+        children2 = sorted(node2.children, key=lambda x: x.name)
+        taxa1 = [x.name for x in children1]
+        taxa2 = [x.name for x in children2]
+        if taxa1 != taxa2:
+            return False
+
+        # recursively test all children
+        res = []
+        for child1, child2 in zip(children1, children2):
+            res.append(compare_topology(child1, child2))
+        if res.count(True) < len(res):
+            return False
+    return True
+
+
+def read_taxdump(nodes_fp, names_fp=None):
+    """Read NCBI taxdump.
+
+    Parameters
+    ----------
+    nodes_fp : str
+        file path to NCBI nodes.dmp
+    names_fp : str, optional
+        file path to NCBI names.dmp
+
+    Returns
+    -------
+    dict of dict
+        taxid : {
+            'parent' : str
+                parent taxid
+            'rank' : str
+                taxonomic rank
+            'name' : str
+                taxon name, empty if names_fp is None
+            'children' : set of str
+                child taxids
+        }
+    """
+    taxdump = {}
+
+    # format of nodes.dmp: taxid | parent taxid | rank | more info...
```

Comment:
I tend to say use Pandas but I remember some odd formatting issues of NCBI taxdumps?!

Reply:
In NCBI taxdump, field delimiters are `\t|\t`, and line ends are `\t|`. I explored pandas' `read_table` but couldn't find an easy way to read it directly.

Reply:
OK